Userland/Libraries/LibGfx/PNGWriter.cpp

```diff
@@ -142,16 +163,91 @@ void PNGWriter::add_IDAT_chunk(Gfx::Bitmap const& bitmap)
+            ByteBuffer buffer {};
+            int sum = 0;
+
+            void append(u8 byte)
+            {
+                buffer.append(byte);
+                sum += static_cast<i8>(byte);
+            }
+
+            void append(AK::SIMD::u8x4 simd)
+            {
+                append(simd[0]);
+                append(simd[1]);
+                append(simd[2]);
+                append(simd[3]);
+            }
+        };
+
+        Filter none_filter { .type = PNG::FilterType::None };
+        none_filter.buffer.ensure_capacity(sizeof(Pixel) * bitmap.height());
+
+        Filter sub_filter { .type = PNG::FilterType::Sub };
+        sub_filter.buffer.ensure_capacity(sizeof(Pixel) * bitmap.height());
+
+        Filter up_filter { .type = PNG::FilterType::Up };
+        up_filter.buffer.ensure_capacity(sizeof(Pixel) * bitmap.height());
+
+        Filter average_filter { .type = PNG::FilterType::Average };
+        average_filter.buffer.ensure_capacity(sizeof(ARGB32) * bitmap.height());
+
+        Filter paeth_filter { .type = PNG::FilterType::Paeth };
+        paeth_filter.buffer.ensure_capacity(sizeof(ARGB32) * bitmap.height());
+
+        auto pixel_x_minus_1 = Pixel::gfx_to_png(*dummy_scanline);
+        auto pixel_xy_minus_1 = Pixel::gfx_to_png(*dummy_scanline);
 
         for (int x = 0; x < bitmap.width(); ++x) {
-            auto pixel = bitmap.get_pixel(x, y);
-            uncompressed_block_data.append(pixel.red());
-            uncompressed_block_data.append(pixel.green());
-            uncompressed_block_data.append(pixel.blue());
-            uncompressed_block_data.append(pixel.alpha());
+            auto pixel = Pixel::gfx_to_png(scanline[x]);
+            auto pixel_y_minus_1 = Pixel::gfx_to_png(scanline_minus_1[x]);
+
+            none_filter.append(pixel);
+
+            sub_filter.append(pixel - pixel_x_minus_1);
+
+            up_filter.append(pixel - pixel_y_minus_1);
+
+            // The sum Orig(a) + Orig(b) shall be performed without overflow (using at least nine-bit arithmetic).
+            auto sum = AK::SIMD::to_u16x4(pixel_x_minus_1) + AK::SIMD::to_u16x4(pixel_y_minus_1);
+            auto average = AK::SIMD::to_u8x4(sum / 2);
+            average_filter.append(pixel - average);
+
+            paeth_filter.append(pixel - PNG::paeth_predictor(pixel_x_minus_1, pixel_y_minus_1, pixel_xy_minus_1));
+
+            pixel_x_minus_1 = pixel;
+            pixel_xy_minus_1 = pixel_y_minus_1;
         }
+
+        scanline_minus_1 = scanline;
+
+        // 12.8 Filter selection: https://www.w3.org/TR/PNG/#12Filter-selection
+        // For best compression of truecolour and greyscale images, the recommended approach
+        // is adaptive filtering in which a filter is chosen for each scanline.
+        // The following simple heuristic has performed well in early tests:
+        // compute the output scanline using all five filters, and select the filter that gives the smallest sum of absolute values of outputs.
```

Comment:
That's a pretty clever technique for guessing which buffer will compress best. :)